File: sn_tool.cpp
#include <windows.h>
#include <wincrypt.h>
#include <stdio.h>
#include <string.h>
#include <string>
#include <vector>
#include <iostream>

static std::string Sha256(const std::string &s)
{
    HCRYPTPROV hProv = 0; HCRYPTHASH hHash = 0; BYTE rgbHash[32]; DWORD cbHash = 32;
    CryptAcquireContext(&hProv, NULL, NULL, PROV_RSA_AES, CRYPT_VERIFYCONTEXT);
    CryptCreateHash(hProv, CALG_SHA_256, 0, 0, &hHash);
    CryptHashData(hHash, (BYTE*)s.data(), (DWORD)s.size(), 0);
    CryptGetHashParam(hHash, HP_HASHVAL, rgbHash, &cbHash, 0);
    CryptDestroyHash(hHash); CryptReleaseContext(hProv, 0);
    return std::string((char*)rgbHash, (char*)rgbHash + 32);
}

static std::string Pkcs7(const std::string &in, size_t bs)
{
    size_t pad = bs - (in.size() % bs);
    if (pad == 0) pad = bs;
    std::string out = in; out.resize(in.size() + pad, (char)pad);
    return out;
}

static std::string Base64Encode(const std::string &bin)
{
    DWORD outLen = 0; CryptBinaryToStringA((BYTE*)bin.data(), (DWORD)bin.size(), CRYPT_STRING_BASE64 | CRYPT_STRING_NOCRLF, NULL, &outLen);
    std::string out; out.resize(outLen);
    CryptBinaryToStringA((BYTE*)bin.data(), (DWORD)bin.size(), CRYPT_STRING_BASE64 | CRYPT_STRING_NOCRLF, &out[0], &outLen);
    if (!out.empty() && out.back() == '\0') out.pop_back();
    return out;
}

static std::string AesCbcEncrypt(const std::string &plain, const std::string &key, const std::string &iv)
{
    HCRYPTPROV hProv = 0; HCRYPTKEY hKey = 0; HCRYPTHASH hHash = 0;
    CryptAcquireContext(&hProv, NULL, NULL, PROV_RSA_AES, CRYPT_VERIFYCONTEXT);
    CryptCreateHash(hProv, CALG_SHA_256, 0, 0, &hHash);
    CryptHashData(hHash, (BYTE*)key.data(), (DWORD)key.size(), 0);
    BYTE raw[32]; DWORD rawLen = 32; CryptGetHashParam(hHash, HP_HASHVAL, raw, &rawLen, 0);
    CryptDestroyHash(hHash);

    struct { BLOBHEADER hdr; DWORD len; BYTE key[32]; } blob;
    blob.hdr.bType = PLAINTEXTKEYBLOB; blob.hdr.bVersion = CUR_BLOB_VERSION; blob.hdr.reserved = 0; blob.hdr.aiKeyAlg = CALG_AES_256;
    blob.len = 32; memcpy(blob.key, raw, 32);
    CryptImportKey(hProv, (BYTE*)&blob, sizeof(blob), 0, 0, &hKey);

    DWORD mode = CRYPT_MODE_CBC; CryptSetKeyParam(hKey, KP_MODE, (BYTE*)&mode, 0);
    BYTE ivBuf[16]; memcpy(ivBuf, iv.data(), 16); CryptSetKeyParam(hKey, KP_IV, ivBuf, 0);

    std::string data = Pkcs7(plain, 16);
    std::string ct = data; DWORD ctLen = (DWORD)ct.size();
    CryptEncrypt(hKey, 0, TRUE, 0, (BYTE*)&ct[0], &ctLen, (DWORD)ct.size());
    CryptDestroyKey(hKey); CryptReleaseContext(hProv, 0);
    return ct;
}

static char PickAlphaNum(const std::string &s, int startIdx)
{
    int n = (int)s.size();
    for (int k = 0; k < n; ++k) {
        char ch = s[(startIdx + k) % n];
        if ((ch >= '0' && ch <= '9') || (ch >= 'A' && ch <= 'Z') || (ch >= 'a' && ch <= 'z')) return ch;
    }
    return '0';
}

static std::string BuildFirst10(const std::string &t, const std::string &m, int sn, std::string &err)
{
    int remain = 10 - ((int)t.size() + (int)m.size());
    if (remain < 1) { err = "前10位长度不合法：产品类型+型号总长度需小于10"; return ""; }
    char buf[64]; sprintf(buf, "%0*d", remain, sn);
    return t + m + std::string(buf);
}

struct Row { int idx; std::string type, model, f10, tail6, did; int sn; };

static std::vector<Row> GenerateRows(const std::string &type, const std::string &model, int snStart, int count, std::string &err)
{
    std::vector<Row> rows; rows.reserve(count);
    for (int i = 0; i < count; ++i) {
        int sn = snStart + i;
        std::string f10 = BuildFirst10(type, model, sn, err);
        if (!err.empty()) return rows;
        std::string key = "123456";
        std::string iv = Sha256(f10).substr(0, 16);
        std::string ct = AesCbcEncrypt(f10, key, iv);
        std::string b64 = Base64Encode(ct);
        int pos[6] = {1,3,6,10,15,21};
        std::string tail(6, '0');
        for (int j = 0; j < 6; ++j) tail[j] = PickAlphaNum(b64, pos[j]-1);
        for (auto &c : tail) c = (char)toupper((unsigned char)c);
        std::string did = f10 + tail;
        rows.push_back({ i+1, type, model, f10, tail, did, sn });
    }
    return rows;
}

static bool WriteXls(const std::vector<Row> &rows, const std::string &name)
{
    std::string fname = name.empty() ? "did_list.xls" : name;
    FILE *fp = fopen(fname.c_str(), "wb"); if (!fp) return false;
    fputs("<!DOCTYPE html><html><head><meta charset=\"utf-8\"></head><body><table border=\"1\"><thead><tr>", fp);
    const char *head[7] = {"#","产品类型","型号","SN","First10","尾6","DID"};
    for (int i = 0; i < 7; ++i) { fprintf(fp, "<th>%s</th>", head[i]); }
    fputs("</tr></thead><tbody>", fp);
    for (auto &r : rows) {
        fprintf(fp, "<tr><td>%d</td><td>%s</td><td>%s</td><td>%d</td><td>%s</td><td>%s</td><td>%s</td></tr>", r.idx, r.type.c_str(), r.model.c_str(), r.sn, r.f10.c_str(), r.tail6.c_str(), r.did.c_str());
    }
    fputs("</tbody></table></body></html>", fp);
    fclose(fp); return true;
}

int main(int argc, char **argv)
{
    std::string type="", model="", out="did_list.xls"; int snStart=1, count=100, preview=10;
    for (int i = 1; i < argc; ++i) {
        std::string a = argv[i];
        if (a == "-type" && i+1<argc) { type = argv[++i]; }
        else if (a == "-model" && i+1<argc) { model = argv[++i]; }
        else if (a == "-snStart" && i+1<argc) { snStart = atoi(argv[++i]); }
        else if (a == "-count" && i+1<argc) { count = atoi(argv[++i]); }
        else if (a == "-out" && i+1<argc) { out = argv[++i]; }
        else if (a == "-preview" && i+1<argc) { preview = atoi(argv[++i]); }
    }
    std::string err;
    auto rows = GenerateRows(type, model, snStart, count, err);
    if (!err.empty()) { fprintf(stderr, "%s\n", err.c_str()); return 1; }
    if (preview > (int)rows.size()) preview = (int)rows.size();
    printf("#\t类型\t型号\tSN\tFirst10\t尾6\tDID\n");
    for (int i = 0; i < preview; ++i) {
        auto &r = rows[i];
        printf("%d\t%s\t%s\t%d\t%s\t%s\t%s\n", r.idx, r.type.c_str(), r.model.c_str(), r.sn, r.f10.c_str(), r.tail6.c_str(), r.did.c_str());
    }
    if (!WriteXls(rows, out)) { fprintf(stderr, "导出失败: %s\n", out.c_str()); return 2; }
    printf("导出完成: %s\n", out.c_str());
    return 0;
}

File: sn_tool.go
package main

import (
    "bufio"
    "crypto/aes"
    "crypto/cipher"
    "crypto/sha256"
    "encoding/base64"
    "flag"
    "fmt"
    "os"
    "strconv"
    "strings"
)

func sha256Bytes(s string) []byte {
    h := sha256.Sum256([]byte(s))
    return h[:]
}

func pkcs7(data []byte, bs int) []byte {
    pad := bs - (len(data) % bs)
    if pad == 0 {
        pad = bs
    }
    out := make([]byte, len(data)+pad)
    copy(out, data)
    for i := len(data); i < len(out); i++ {
        out[i] = byte(pad)
    }
    return out
}

func ab2b64(b []byte) string {
    return base64.StdEncoding.EncodeToString(b)
}

func pickAlphaNum(s string, startIdx int) byte {
    n := len(s)
    for k := 0; k < n; k++ {
        ch := s[(startIdx+k)%n]
        if (ch >= '0' && ch <= '9') || (ch >= 'A' && ch <= 'Z') || (ch >= 'a' && ch <= 'z') {
            return ch
        }
    }
    return '0'
}

func encTail(first10 string) (string, error) {
    keyBytes := sha256Bytes("123456")
    block, err := aes.NewCipher(keyBytes)
    if err != nil {
        return "", err
    }
    iv := sha256Bytes(first10)[:16]
    data := pkcs7([]byte(first10), block.BlockSize())
    cipherText := make([]byte, len(data))
    cipher.NewCBCEncrypter(block, iv).CryptBlocks(cipherText, data)
    b64 := ab2b64(cipherText)
    positions := []int{1, 3, 6, 10, 15, 21}
    out := make([]byte, 6)
    for i, p := range positions {
        out[i] = pickAlphaNum(b64, p-1)
    }
    return strings.ToUpper(string(out)), nil
}

func buildFirst10(t, m string, sn int) (string, error) {
    tl := len(t)
    ml := len(m)
    remain := 10 - (tl + ml)
    if remain < 1 {
        return "", fmt.Errorf("前10位长度不合法：产品类型+型号总长度需小于10")
    }
    snStr := fmt.Sprintf("%0*d", remain, sn)
    return t + m + snStr, nil
}

type Row struct {
    Idx   int
    Type  string
    Model string
    SN    int
    F10   string
    Tail6 string
    DID   string
}

func generateRows(t, m string, snStart, count int) ([]Row, error) {
    rows := make([]Row, 0, count)
    for i := 0; i < count; i++ {
        sn := snStart + i
        f10, err := buildFirst10(t, m, sn)
        if err != nil {
            return rows, err
        }
        tail6, err := encTail(f10)
        if err != nil {
            return rows, err
        }
        did := f10 + tail6
        rows = append(rows, Row{Idx: i + 1, Type: t, Model: m, SN: sn, F10: f10, Tail6: tail6, DID: did})
    }
    return rows, nil
}

func writeXls(rows []Row, name string) error {
    if name == "" {
        name = "did_list.xls"
    }
    f, err := os.Create(name)
    if err != nil {
        return err
    }
    defer f.Close()
    w := bufio.NewWriter(f)
    defer w.Flush()
    head := []string{"#", "产品类型", "型号", "SN", "First10", "尾6", "DID"}
    fmt.Fprintf(w, "<!DOCTYPE html><html><head><meta charset=\"utf-8\"></head><body><table border=\"1\"><thead><tr>")
    for _, h := range head {
        fmt.Fprintf(w, "<th>%s</th>", h)
    }
    fmt.Fprintf(w, "</tr></thead><tbody>")
    for _, r := range rows {
        fmt.Fprintf(w, "<tr>")
        fmt.Fprintf(w, "<td>%d</td>", r.Idx)
        fmt.Fprintf(w, "<td>%s</td>", r.Type)
        fmt.Fprintf(w, "<td>%s</td>", r.Model)
        fmt.Fprintf(w, "<td>%d</td>", r.SN)
        fmt.Fprintf(w, "<td>%s</td>", r.F10)
        fmt.Fprintf(w, "<td>%s</td>", r.Tail6)
        fmt.Fprintf(w, "<td>%s</td>", r.DID)
        fmt.Fprintf(w, "</tr>")
    }
    fmt.Fprintf(w, "</tbody></table></body></html>")
    return nil
}

func main() {
    t := flag.String("type", "", "产品类型，如 PT")
    m := flag.String("model", "", "型号，如 MD100")
    snStart := flag.Int("snStart", 1, "起始SN")
    count := flag.Int("count", 1, "生成数量")
    file := flag.String("out", "did_list.xls", "导出文件名")
    preview := flag.Int("preview", 10, "控制台预览条数")
    flag.Parse()

    rows, err := generateRows(*t, *m, *snStart, *count)
    if err != nil {
        fmt.Fprintln(os.Stderr, err)
        os.Exit(1)
    }
    n := *preview
    if n > len(rows) {
        n = len(rows)
    }
    fmt.Println("#\t类型\t型号\tSN\tFirst10\t尾6\tDID")
    for i := 0; i < n; i++ {
        r := rows[i]
        fmt.Println(strconv.Itoa(r.Idx) + "\t" + r.Type + "\t" + r.Model + "\t" + strconv.Itoa(r.SN) + "\t" + r.F10 + "\t" + r.Tail6 + "\t" + r.DID)
    }
    if err := writeXls(rows, *file); err != nil {
        fmt.Fprintln(os.Stderr, err)
        os.Exit(2)
    }
    fmt.Println("导出完成:", *file)
}
